src/registers/debug.rs

```diff
@@ -0,0 +1,603 @@
+
+    /// 2 byte length
+    Length2B = 0b01,
+
+    /// 8 byte length
+    Length8B = 0b10,
+
+    /// 4 byte length
+    Length4B = 0b11,
+}
+
+impl HwBreakpointSize {
+    /// Creates a new hardware breakpoint size if `size` is valid.
+    pub const fn new(size: usize) -> Option<Self> {
+        match size {
+            1 => Some(Self::Length1B),
+            2 => Some(Self::Length2B),
+            8 => Some(Self::Length8B),
+            4 => Some(Self::Length4B),
+            _ => None,
+        }
+    }
+
+    const fn bit_position(n: DebugAddressRegisterNumber) -> u8 {
+        18 + 4 * n.0
+    }
+}
+
+impl TryFrom<usize> for HwBreakpointSize {
+    type Error = TryFromIntError;
+
+    fn try_from(size: usize) -> Result<Self, Self::Error> {
+        Self::new(size).ok_or(TryFromIntError(()))
+    }
+}
+
+/// A valid value of the [`Dr7`] debug register.
+///
+/// In addition to the [`Dr7Flags`] this value has a condition field and a size field for each debug address register.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct Dr7Value {
+    bits: u64,
+}
+
+impl From<Dr7Flags> for Dr7Value {
+    fn from(dr7_flags: Dr7Flags) -> Self {
+        Self::from_bits_truncate(dr7_flags.bits())
+    }
+}
+
+impl Dr7Value {
+    const fn mask() -> u64 {
+        let field_mask = (1 << 32) - (1 << 16);
+        let flag_mask = Dr7Flags::all().bits();
+        field_mask | flag_mask
+    }
+
+    /// Convert from underlying bit representation, unless that representation contains bits that do not correspond to a field.
+    #[inline]
+    pub const fn from_bits(bits: u64) -> Option<Self> {
+        if (bits & Self::mask()) == 0 {
+            Some(Self { bits })
+        } else {
+            None
+        }
+    }
+
+    /// Convert from underlying bit representation, dropping any bits that do not correspond to fields.
+    #[inline]
+    pub const fn from_bits_truncate(bits: u64) -> Self {
+        Self {
+            bits: bits & !Self::mask(),
+        }
+    }
+
+    /// Convert from underlying bit representation, preserving all bits (even those not corresponding to a defined field).
+    ///
+    /// # Safety
+    ///
+    /// The bit representation must be a valid [`Dr7Value`].
+    #[inline]
+    pub const unsafe fn from_bits_unchecked(bits: u64) -> Self {
+        Self { bits }
+    }
+
+    /// Returns the raw value of the fields currently stored.
+    #[inline]
+    pub const fn bits(&self) -> u64 {
+        self.bits
+    }
+
+    /// Returns the [`Dr7Flags`] in this value.
+    #[inline]
+    pub const fn flags(self) -> Dr7Flags {
+        // SAFETY: The extra bits used in fields are valid in our representation of `Dr7Flags`.
+        unsafe { Dr7Flags::from_bits_unchecked(self.bits) }
+    }
+
+    /// Returns a mutable reference to the [`Dr7Flags`] in this value.
+    ///
+    /// This can be used to modify the flags of this value in-place.
+    #[inline]
+    pub fn flags_mut(&mut self) -> &mut Dr7Flags {
```

Comment:
I'm not completely happy with this function, as it could be misused.

I added it to have a way of modifying the `Dr7Flags` bits inside the `Dr7Value`. Doing so via this mutable reference causes unintended side effects to the fields in `Dr7Value`. Perhaps it would be better to manually implement all operations for adding/removing flags from the value instead of transmuting self to the flags. What do you think?

Reply:
If additional bits are allowed for `Dr7Flags`, can't we make use `Dr7Flags` for the inner value directly (instead of `u64`)?

Alternatively, we could also merge the `Dr7Value` type into the `Dr7Flags` type by defining the additional methods in an `impl Dr7Flags` block. To get the same behavior for `from_bits_truncate`, we could define an addional bitflag const for all the additional value, i.e. `const CONDITIONS_AND_SIZE = (1 << 32) - (1 << 16);`.

Reply:
Ah, yes. We could use `Dr7Flags` for the inner value.

It would expose the same (problematic?) behavior as defining methods on `Dr7Flags` or the current behavior: When you have a valid `Dr7Value`, there are values which are no bitflags (`HwBreakpointCondition` and `HwBreakpointSize`). If those values are set to some non-zero value, the behavior of `BitAnd` and similar traits are not behaving as one might expect.

So I think to cleanly solve this problem, we would have to remove `flags_mut` and implement `BitAnd<Flags> for Value` and similar traits manually. We could additionally make use of `Dr7Flags` for the inner value, but I am not sure yet if that would be beneficial. What do you think?

Reply:
All right. I went ahead and removed this method and any unsafe code handling this type in https://github.com/rust-osdev/x86_64/pull/286/commits/0eff064881570b0ec0c67821e0495ac81b0f2288.

I forbid any use of extra bits in the representation of `Dr7Flags` and added the explicit `*_flags` methods for manipulating `Dr7Value`. What do you think?